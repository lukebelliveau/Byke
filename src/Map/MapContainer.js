import { connect } from 'react-redux';
import MapComponent from './Map';
import actions from '../redux/actions';

const mapDispatchToProps = dispatch => ({
  startLoading: () => {
    dispatch(actions.loadingStarted());
  },
});

const Map = connect(mapDispatchToProps)(MapComponent);

export default Map;
