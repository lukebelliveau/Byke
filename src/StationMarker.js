import React from 'react';
import MapView from 'react-native-maps';

import Circle from './Circle';

const Marker = ({ coordinate, stationName, availableBikes }) =>
  <MapView.Marker
    coordinate={coordinate}
    title={stationName}
    description={`${availableBikes} bikes available`}
  >
    <Circle color={availableBikes > 5 ? 'green' : 'red'} />
  </MapView.Marker>;

export default Marker;
