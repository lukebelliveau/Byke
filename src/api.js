// @flow
export const url = 'http://byke-graphql.herokuapp.com/graphql?query=';

const googleMapsAPIKey = 'AIzaSyAQhawlRtPkYdpC4VHgaB1n3yLOVOEZbYE';

const getAllStations = (stationParams: string) =>
  fetch(`${url}{allStations${stationParams}}`)
    .then(response => {
      return response.json().then(res => {
        return res.data.allStations;
      });
    })
    .catch(e => console.log(e));

const searchPlaces = (searchQuery: string, latitude, longitude) =>
  fetch(
    `https://maps.googleapis.com/maps/api/place/nearbysearch/json?key=${googleMapsAPIKey}&location=${latitude},${longitude}&radius=10000&keyword=${searchQuery}`
  )
    .then(response => {
      response.json().then(json => json.results);
    })
    .catch(e => console.log(e));

export default {
  getAllStations,
  searchPlaces,
};
