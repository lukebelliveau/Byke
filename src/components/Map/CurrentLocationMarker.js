import React from 'react';
import { View, Keyboard } from 'react-native';
import MapView from 'react-native-maps';

const CurrentLocationMarker = ({ coordinate }) =>
  <MapView.Marker
    coordinate={coordinate}
    title="Your location"
    onPress={Keyboard.dismiss}
  >
    <View style={style} />
  </MapView.Marker>;

const style = {
  width: 20,
  height: 20,
  borderRadius: 50,
  backgroundColor: 'dodgerblue',
  borderColor: 'white',
  borderWidth: 1,
};

export default CurrentLocationMarker;
