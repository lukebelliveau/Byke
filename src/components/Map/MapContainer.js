// @flow
import { connect } from 'react-redux';
import MapComponent from './Map';
import actions from '../../redux/actions';

const mapDispatchToProps = dispatch => ({
  loadingStarted: () => {
    dispatch(actions.loadingStarted());
  },
  loadingFinished: () => {
    dispatch(actions.loadingFinished());
  },
  stationsFetched: stations => {
    dispatch(actions.stationsFetched(stations));
  },
  locationUpdated: location => {
    dispatch(actions.locationUpdated(location));
  },
});

const mapStateToProps = state => ({
  stations: state.stations,
  trip: state.trip,
  currentLocation: state.currentLocation,
  region: state.region,
});

const Map = connect(mapStateToProps, mapDispatchToProps)(MapComponent);

export default Map;
