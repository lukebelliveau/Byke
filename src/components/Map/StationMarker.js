// @flow
import React from 'react';
import MapView from 'react-native-maps';
import { Keyboard } from 'react-native';
import utils from '../../utils';

import Circle from '../shared/Circle';

const Marker = ({
  coordinate,
  stationName,
  currentLocation,
  availableBikes,
  availableDocks,
}: {
  coordinate: Object,
  stationName: string,
  availableBikes: number,
}) => {
  return (
    <MapView.Marker
      coordinate={coordinate}
      title={stationName}
      description={`${availableBikes} bikes available, ${availableDocks} docks available`}
      onPress={Keyboard.dismiss}
      onCalloutPress={() =>
        utils.displayNavigationAlert(
          currentLocation,
          stationName,
          coordinate,
          availableBikes,
          availableDocks
        )}
    >
      <Circle color={availableBikes > 5 ? 'green' : 'red'} />
    </MapView.Marker>
  );
};

export default Marker;
