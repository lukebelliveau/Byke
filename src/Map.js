// @flow
import React, { Component } from 'react';
import { Keyboard, View, AsyncStorage } from 'react-native';
import MapView from 'react-native-maps';

import StationMarker from './StationMarker';
import api from './api';
import Loading from './Loading';

type Station = {
  latitude: number,
  longitude: number,
  stationName: string,
  availableBikes: number,
};

type Props = {
  region: { number: string },
  stations: Array<Station>,
  style: Object,
  trip: Object,
};

const initialState = {
  stations: [],
};
class Map extends Component {
  state = initialState;
  constructor(props: Props) {
    super(props);

    this.props = props;
  }

  componentDidMount() {
    this.fetchStationInfo();
  }

  componentDidUpdate() {
    this.map.animateToRegion(this.props.region)
  }

  loadStations() {
    const stations = AsyncStorage.getItem('@Byke:stations')
      .then(stations => {
        if (stations !== null) {
          this.setState({
            stations: JSON.parse(stations),
          });
        } else {
          this.fetchStationInfo();
        }
      })
      .catch(e => {
        Error('error fetching stations!' + e);
      });
  }

  fetchStationInfo = () => {
    api
      .getAllStations(`{ stationName, availableBikes, latitude, longitude }`)
      .then(stations => {
        AsyncStorage.setItem(
          '@Byke:stations',
          JSON.stringify(stations)
        ).catch(e => console.log('error:' + e));
        this.setState({
          stations,
        });
      });
  };

  render() {
    const trip = this.props.trip;
    const region = this.props.region;
    const stations = this.state.stations;

    return (
      <View style={{ flex: 1 }}>
        {
          <MapView
            style={{ flex: 1 }}
            //region={region}
            initialRegion={region}
            onPress={Keyboard.dismiss}
            ref={ref => { this.map = ref; }}
          >
            {trip
              ? <View>
                  <MapView.Marker
                    coordinate={trip.currentLocation}
                    pinColor="blue"
                  />
                  <MapView.Marker coordinate={trip.destination} />
                </View>
              : <MapView.Marker coordinate={region} />}

            {stations.map((station, index) =>
              <StationMarker
                testId={station.stationName}
                stationName={station.stationName}
                coordinate={station}
                availableBikes={station.availableBikes}
                key={index}
              />
            )}
          </MapView>
        }
        {stations.length === 0 ? <Loading /> : null}
      </View>
    );
  }
}
export default Map;
