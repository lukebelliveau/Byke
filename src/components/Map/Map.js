// @flow
import React, { Component } from 'react';
import { Keyboard, View } from 'react-native';
import MapView from 'react-native-maps';

import StationMarker from './StationMarker';
import CurrentLocationMarker from './CurrentLocationMarker';
import api from '../../api';
import { modes } from '../../redux/reducers';
import { Region, Location, Station, Trip } from '../../Types';

type Props = {
  stations: Array<Station>,
  region: Region,
  trip: Trip,
  loadingStarted: () => void,
  loadingFinished: () => void,
  locationUpdated: Location => void,
  stationsFetched: (Array<Station>) => void,
};

const positionError = error => {
  throw Error(error.message);
};

const positionOptions = {
  timeout: Number.POSITIVE_INFINITY,
  maximumAge: 0,
  enableHighAccuracy: false,
};

class Map extends Component {
  map: {
    animateToRegion: Region => void,
  };
  constructor(props: Props) {
    super(props);

    this.props = props;
    this.fetchStationInfo();
    this.watchLocation();
  }

  watchLocation() {
    navigator.geolocation.watchPosition(
      location => {
        this.props.locationUpdated(location.coords), null, {};
      },
      positionError,
      positionOptions
    );
  }

  componentDidUpdate() {
    this.map.animateToRegion(this.props.region);
  }

  fetchStationInfo = () => {
    this.props.loadingStarted();
    api
      .getAllStations(
        `{ stationName, availableBikes, availableDocks, latitude, longitude }`
      )
      .then(stations => {
        this.props.stationsFetched(stations);
        this.props.loadingFinished();
      });
  };

  render() {
    const trip = this.props.trip;
    const region = this.props.region;
    const stations = this.props.stations;
    const mode = this.props.mode;
    const currentLocation = this.props.currentLocation;

    return (
      <View style={{ flex: 1 }}>
        {
          <MapView
            style={{ flex: 1 }}
            onPress={Keyboard.dismiss}
            ref={ref => {
              this.map = ref;
            }}
          >
            <CurrentLocationMarker coordinate={currentLocation} />

            {mode === modes.tripDisplay
              ? <View>
                  <MapView.Marker coordinate={trip.destination} />
                  {trip.closeToLocation.map(stationIndex => {
                    const station = stations[stationIndex];
                    return (
                      <StationMarker
                        testId={station.stationName}
                        stationName={station.stationName}
                        coordinate={station}
                        currentLocation={currentLocation}
                        availableBikes={station.availableBikes}
                        availableDocks={station.availableDocks}
                        key={stationIndex}
                      />
                    );
                  })}
                </View>
              : null}

            {mode === modes.overview
              ? stations.map((station, index) => {
                  return (
                    <StationMarker
                      testId={station.stationName}
                      stationName={station.stationName}
                      coordinate={station}
                      currentLocation={currentLocation}
                      availableBikes={station.availableBikes}
                      availableDocks={station.availableDocks}
                      key={index}
                    />
                  );
                })
              : null}
          </MapView>
        }
      </View>
    );
  }
}

export default Map;
