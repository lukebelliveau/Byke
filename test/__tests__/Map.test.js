import React from 'react';
import renderer from 'react-test-renderer';
import MapView from 'react-native-maps';
import { shallow } from 'enzyme';
import { Keyboard } from 'react-native';
import { createWaitForElement } from 'enzyme-wait';
import toJson from 'enzyme-to-json';

import Map from '../../src/Map';
import StationMarker from '../../src/StationMarker';
import api from '../../src/api';

const region = {
  latitude: 1,
  longitude: 2,
  latitudeDelta: 3,
  longitudeDelta: 4,
};
const stations = [
  { stationName: 'station1', availableBikes: 1, latitude: 50, longitude: 60 },
];
const trip = {
  currentLocation: { latitude: 99, longitude: 99 },
  destination: { latitude: 199, longitude: 199 },
};

api.getAllStations = jest.fn(() => new Promise(resolve => resolve(stations)));

it('fetches stations and displays them', done => {
  const waitForStations = createWaitForElement(StationMarker);
  const map = shallow(<Map region={region} />);

  map.instance().componentDidMount();

  waitForStations(map).then(component => {
    expect(toJson(component)).toMatchSnapshot();
    done();
  });
});

it('renders a trip with marker in location and destination', () => {
  const map = renderer.create(<Map region={region} trip={trip} />).toJSON();

  expect(map).toMatchSnapshot();
});

it('dismisses the keyboard when pressed', () => {
  Keyboard.dismiss = jest.fn();
  const map = shallow(<Map region={region} stations={stations} style={{}} />);

  map.find(MapView).simulate('press');

  expect(Keyboard.dismiss.mock.calls.length).toBeGreaterThan(0);
});

/*
* CONFIG (react-test-renderer & react-native-maps don't play nice)
* see https://github.com/airbnb/react-native-maps/issues/889
* */
jest.mock('react-native-maps', () => {
  const React = require.requireActual('react');
  const MapView = require.requireActual('react-native-maps');

  class MockCallout extends React.Component {
    render() {
      return React.createElement('Callout', this.props, this.props.children);
    }
  }

  class MockMarker extends React.Component {
    render() {
      return React.createElement('Marker', this.props, this.props.children);
    }
  }

  class MockMapView extends React.Component {
    render() {
      return React.createElement('MapView', this.props, this.props.children);
    }
  }

  MockCallout.propTypes = MapView.Callout.propTypes;
  MockMarker.propTypes = MapView.Marker.propTypes;
  MockMapView.propTypes = MapView.propTypes;
  MockMapView.Marker = MockMarker;
  MockMapView.Callout = MockCallout;
  return MockMapView;
});
