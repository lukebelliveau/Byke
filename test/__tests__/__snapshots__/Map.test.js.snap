// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays stations from state 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <MapView
    onPress={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Marker
      coordinate={
        Object {
          "latitude": 50,
          "longitude": 50,
        }
      }
      onPress={[Function]}
      title="Your location"
    >
      <View
        style={
          Object {
            "backgroundColor": "dodgerblue",
            "borderColor": "white",
            "borderRadius": 50,
            "borderWidth": 1,
            "height": 20,
            "width": 20,
          }
        }
      />
    </Marker>
    <Marker
      coordinate={
        Object {
          "availableBikes": 1,
          "availableDocks": 10,
          "latitude": 50,
          "longitude": 60,
          "stationName": "station1",
        }
      }
      description="1 bikes available, 10 docks available"
      onCalloutPress={[Function]}
      onPress={[Function]}
      title="station1"
    >
      <View
        style={
          Object {
            "backgroundColor": "red",
            "borderRadius": 10,
            "height": 15,
            "width": 15,
          }
        }
      />
    </Marker>
  </MapView>
</View>
`;

exports[`renders a trip with marker in location and destination 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <MapView
    onPress={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Marker
      coordinate={
        Object {
          "latitude": 50,
          "longitude": 50,
        }
      }
      onPress={[Function]}
      title="Your location"
    >
      <View
        style={
          Object {
            "backgroundColor": "dodgerblue",
            "borderColor": "white",
            "borderRadius": 50,
            "borderWidth": 1,
            "height": 20,
            "width": 20,
          }
        }
      />
    </Marker>
    <View>
      <Marker
        coordinate={
          Object {
            "latitude": 199,
            "longitude": 199,
          }
        }
      />
      <Marker
        coordinate={
          Object {
            "availableBikes": 1,
            "availableDocks": 10,
            "latitude": 50,
            "longitude": 60,
            "stationName": "station1",
          }
        }
        description="1 bikes available, 10 docks available"
        onCalloutPress={[Function]}
        onPress={[Function]}
        title="station1"
      >
        <View
          style={
            Object {
              "backgroundColor": "red",
              "borderRadius": 10,
              "height": 15,
              "width": 15,
            }
          }
        />
      </Marker>
    </View>
  </MapView>
</View>
`;
